closed/src/jdk.crypto.ec/share/classes/sun/security/ec/NativeXDHKeyAgreement.java

```diff
@@ -0,0 +1,320 @@
+
+            // Ensure public key parameters are compatible with private key
+            XECParameters xecParams = XECParameters.get(InvalidKeyException::new,
+                publicKey.getParams());
+            if (!ops.getParameters().oidEquals(xecParams)) {
+                throw new InvalidKeyException(
+                    "Public key parameters are not compatible with private key.");
+            }
+
+            /* Assign ID used by OpenSSL for different curves */
+            int curveType;
+            if ("X25519".equalsIgnoreCase(ops.getParameters().getName())) {
+                curveType = NativeCrypto.X25519;
+            } else {
+                curveType = NativeCrypto.X448;
+            }
+
+            byte[] publicKeyArray = publicKey.getKeyAsByteArray();
+            if (publicKeyArray == null) {
+                throw new InvalidKeyException("No public key value");
+            }
+            computedSecret = new byte[ops.getParameters().getBytes()];
+
+            if (nativeCrypto == null) {
+                nativeCrypto = NativeCrypto.getNativeCrypto();
+            }
+            int result = nativeCrypto.XDHGenerateSecret(privateKey, privateKey.length,
+                                                        publicKeyArray, publicKeyArray.length,
+                                                        computedSecret, computedSecret.length,
+                                                        curveType);
+
+            if (result == -1) {
+                if (nativeCryptTrace) {
+                    System.err.println("Shared secret generation by OpenSSL failed," +
+                        " using Java crypto implementation.");
+                }
+                computedSecret = javaImplComputeSecret(key, lastPhase);
+            }
+        }
+
+        // test for contributory behavior
+        if (allZero(computedSecret)) {
+            throw new InvalidKeyException("Point has small order");
+        }
+
+        this.secret = computedSecret;
+        return null;
+    }
+
+    /*
+     * Constant-time check for an all-zero array.
+     */
+    private static boolean allZero(byte[] arr) {
+        byte orValue = (byte) 0;
+        for (int i = 0; i < arr.length; i++) {
+            orValue |= arr[i];
+        }
+
+        return orValue == (byte) 0;
+    }
+
+    @Override
+    protected byte[] engineGenerateSecret() throws IllegalStateException {
+        byte[] result = secret;
+        if (result == null) {
+            throw new IllegalStateException("Not initialized correctly");
+        }
+        secret = null;
+        return result;
+    }
+
+    @Override
+    protected int engineGenerateSecret(byte[] sharedSecret, int offset)
+        throws IllegalStateException, ShortBufferException {
+
+        if (secret == null) {
+            throw new IllegalStateException("Not initialized correctly");
+        }
+        int secretLen = this.secret.length;
```

Comment:
Please remove `this.` qualification (it wasn't necessary on line 243, so it's not needed here).
Please apply the same reasoning throughout this change.